### PR TITLE
#897: fix to crash because of bug in Substitute for ForallStmt

### LINKED ISSUE #897: Crash with forall statement
In Dafny 3.0.0pre1 (on mac):

```dafny
function missing_number(nums: seq<nat>): nat
    requires |set i | i in nums| == |nums|
    requires forall x :: x in nums ==> 0 <= x <= |nums|
{
    var p := x => 0 <= x <= |nums| && x !in nums;
    assert exists x {:trigger p(x)} :: p(x) && forall y {:trigger p(y)} :: p(y) ==> y == x by {
        var range := set i | 0 <= i <= |nums|;
        assert |range| == |nums| + 1;
        var missing := range - set i | i in nums;
        assert |missing| == 1;
        var x :| x in missing;
        assert p(x);
        forall y | p(y) ensures y == x {}
    }
    var x :| p(x);
    x
}
```

The output is:

```console
$ dafny ./Bug.dfy | cat
Dafny 3.0.0.20820
./Bug.dfy(7,21): Warning: /!\ No terms found to trigger on.
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.Dafny.Translator.TrForallProof (Microsoft.Dafny.ForallStmt s, Microsoft.Dafny.Translator+BoogieStmtListBuilder definedness, Microsoft.Dafny.Translator+BoogieStmtListBuilder exporter, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x00234] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.TrStmt (Microsoft.Dafny.Statement stmt, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x02366] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.TrStmtList (System.Collections.Generic.List`1[T] stmts, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x0009e] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.TrStmt (Microsoft.Dafny.Statement stmt, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x018c2] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.TrStmt (Microsoft.Dafny.Statement stmt, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x00194] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.CheckWellformedWithResult (Microsoft.Dafny.Expression expr, Microsoft.Dafny.Translator+WFOptions options, Microsoft.Boogie.Expr result, Microsoft.Dafny.Type resultType, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x036a1] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.CheckWellformedLetExprWithResult (Microsoft.Dafny.LetExpr e, Microsoft.Dafny.Translator+WFOptions options, Microsoft.Boogie.Expr result, Microsoft.Dafny.Type resultType, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, Microsoft.Dafny.Translator+ExpressionTranslator etran, System.Boolean checkRhs) [0x0017e] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.CheckWellformedWithResult (Microsoft.Dafny.Expression expr, Microsoft.Dafny.Translator+WFOptions options, Microsoft.Boogie.Expr result, Microsoft.Dafny.Type resultType, System.Collections.Generic.List`1[T] locals, Microsoft.Dafny.Translator+BoogieStmtListBuilder builder, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x030ea] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.AddWellformednessCheck (Microsoft.Dafny.Function f) [0x00ae2] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.AddFunction_Top (Microsoft.Dafny.Function f) [0x0003f] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.AddClassMembers (Microsoft.Dafny.TopLevelDeclWithMembers c, System.Boolean includeMethods) [0x0038a] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.AddTypeDecl (Microsoft.Dafny.RevealableTypeDecl d) [0x0005a] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator.DoTranslation (Microsoft.Dafny.Program p, Microsoft.Dafny.ModuleDefinition forModule) [0x00437] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.Translator+<Translate>d__47.MoveNext () [0x000ea] in <dfa96c1a805442f8a00af8d849665edd>:0 
  at Microsoft.Dafny.DafnyDriver+<Translate>d__7.MoveNext () [0x00119] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0012f] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0004b] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <6fd7bea9fd0c48cc9a4e825f7a106a9b>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <6ae2a9bbfe38412a965297a6b0a61fc2>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <6ae2a9bbfe38412a965297a6b0a61fc2>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <6ae2a9bbfe38412a965297a6b0a61fc2>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <6ae2a9bbfe38412a965297a6b0a61fc2>:0 
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <6ae2a9bbfe38412a965297a6b0a61fc2>:0 
```

Changing the `forall` statement to an `assert forall y ...;` avoids this issue. (It doesn't verify successfully, as Dafny isn't quite sold on all the other steps in this proof, but that's my problem!)

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #897.

### CODE REVIEW

File: Test/git-issues/git-issue-897a.dfy [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+function missing_number(nums: seq<nat>): nat
+  requires |set i | i in nums| == |nums|
+  requires forall x :: x in nums ==> 0 <= x <= |nums|
+{
+  var p := x => 0 <= x <= |nums| && x !in nums;
+  assert exists x {:trigger p(x)} :: p(x) && forall y {:trigger p(y)} :: p(y) ==> y == x by {
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-897a.dfy:9]:
```suggestion
  assert exists x :: p(x) && forall y :: p(y) ==> y == x by {
```

File: Test/git-issues/git-issue-897.dfy [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+function missing_number(nums: seq<nat>): nat
+  requires |set i | i in nums| == |nums|
+  requires forall x :: x in nums ==> 0 <= x <= |nums|
+{
+  var p := x => 0 <= x <= |nums| && x !in nums;
+  assert exists x {:trigger p(x)} :: p(x) && forall y {:trigger p(y)} :: p(y) ==> y == x by {
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-897.dfy:9]:
```suggestion
  assert exists x :: p(x) && forall y :: p(y) ==> y == x by {
```

I view `:trigger` as an advanced feature that shouldn't need to be present in user code. In any case, Dafny picks these triggers for this example, so there's no use in listing them explicitly.

File: Source/Dafny/Translator.cs [hunk 3]
```diff
@@ -18735,9 +18735,19 @@ public class Substituter
           foreach (var bv in s.BoundVars) {
             substMap.Remove(bv);
           }
+
           // Put things together
           var rr = new ForallStmt(s.Tok, s.EndTok, newBoundVars, SubstAttributes(s.Attributes), Substitute(s.Range), s.Ens.ConvertAll(SubstMayBeFreeExpr), body);
           rr.Kind = s.Kind;
+          rr.CanConvert = s.CanConvert;
+          rr.Bounds = s.Bounds; // not substituted (should it be?)
```
Comment (RustanLeino) [inline Source/Dafny/Translator.cs:18743]: Yes, I think it should be (by calling `SubstituteBoundedPoolList`). The bounds may contain the variables that are being substituted.

File: Source/Dafny/Translator.cs [hunk 4]
```diff
@@ -18735,9 +18735,19 @@ public class Substituter
...
+          if (s.ForallExpressions != null) {
+            var list = new List<Expression>();
+            foreach (Expression e in s.ForallExpressions) {
+              list.Add(Substitute(e));
+            }
+            rr.ForallExpressions = list;
```
Comment (RustanLeino) [inline Source/Dafny/Translator.cs:18749]: A more concise way to write this is

    rr.ForallExpression.ConvertAll(Substitute);

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Test/git-issues/git-issue-897a.dfy (hunk 1 above)
Comment (davidcok) [reply Test/git-issues/git-issue-897a.dfy:9]: OK

File: Source/Dafny/Translator.cs (hunk 3 above)
Comment (davidcok) [reply Source/Dafny/Translator.cs:18743]: OK

File: Source/Dafny/Translator.cs (hunk 4 above)
Comment (davidcok) [reply Source/Dafny/Translator.cs:18749]: OK

File: Test/git-issues/git-issue-897.dfy (hunk 2 above)
Comment (davidcok) [reply Test/git-issues/git-issue-897.dfy:9]: OK

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**